nav2_ceres_costaware_smoother/include/nav2_ceres_costaware_smoother/options.hpp

```diff
@@ -0,0 +1,169 @@
+
+/**
+ * @struct nav2_smac_planner::SmootherParams
+ * @brief Parameters for the smoother cost function
+ */
+struct SmootherParams
+{
+  /**
+   * @brief A constructor for nav2_smac_planner::SmootherParams
+   */
+  SmootherParams()
+  {
+  }
+
+  /**
+   * @brief Get params from ROS parameter
+   * @param node_ Ptr to node
+   * @param name Name of plugin
+   */
+  void get(rclcpp_lifecycle::LifecycleNode * node, const std::string & name)
+  {
+    std::string local_name = name + std::string(".");
+
+    // Smoother params
+    double minimum_turning_radius;
+    nav2_util::declare_parameter_if_not_declared(
+      node, name + ".minimum_turning_radius", rclcpp::ParameterValue(0.4));
+    node->get_parameter(name + ".minimum_turning_radius", minimum_turning_radius);
+    max_curvature = 1.0f / minimum_turning_radius;
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "w_curve", rclcpp::ParameterValue(30.0));
+    node->get_parameter(local_name + "w_curve", curvature_weight);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "w_cost", rclcpp::ParameterValue(0.015));
+    node->get_parameter(local_name + "w_cost", costmap_weight);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "w_cost_cusp", rclcpp::ParameterValue(0.04));
+    node->get_parameter(local_name + "w_cost_cusp", cusp_costmap_weight);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "cusp_zone_length", rclcpp::ParameterValue(-1.0));
+    node->get_parameter(local_name + "cusp_zone_length", cusp_zone_length);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "w_dist", rclcpp::ParameterValue(0.0));
+    node->get_parameter(local_name + "w_dist", distance_weight);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "w_smooth", rclcpp::ParameterValue(15000.0));
+    node->get_parameter(local_name + "w_smooth", smooth_weight);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "cost_check_points", rclcpp::ParameterValue(std::vector<double>()));
+    node->get_parameter(local_name + "cost_check_points", cost_check_points);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "input_downsampling_factor", rclcpp::ParameterValue(1));
+    node->get_parameter(local_name + "input_downsampling_factor", input_downsampling_factor);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "output_upsampling_factor", rclcpp::ParameterValue(1));
+    node->get_parameter(local_name + "output_upsampling_factor", output_upsampling_factor);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "reversing_enabled", rclcpp::ParameterValue(true));
+    node->get_parameter(local_name + "reversing_enabled", reversing_enabled);
+    nav2_util::declare_parameter_if_not_declared(
+      node, local_name + "keep_goal_orientation", rclcpp::ParameterValue(true));
```

Comment:
What situation would we not want to keep the goal / start orientation?

Reply:
e.g. when smoothing a path from SmacPlanner2D when goal checker's `goal_tolerance_yaw` is `.inf`. I.e. when we don't care about current nor final robot orientation, we just want the path to be smooth and far from obstacles. Sticking on the orientations can introduce minor artifacts to final path.

Reply:
Would it be reasonable to restrict this smoother to only kinematically aware plans, given the curvature constraint term? If a plan was generated via another method, I'm not entirely sure how reliably this could / would create a path that followed that constraint if it is wildly deviating from it in the first place (or in some cases where the path itself went through a space where it is impossible to have the curvature constraint followed).

If it can be applied elsewhere, I think that's a reasonable point to make -- but for feasible paths its not possible to create said path without orientations being meaningful for obvious reasons. 

> when smoothing a path from SmacPlanner2D... Sticking on the orientations can introduce minor artifacts to final path.

This would be a good point to make in documentation

Reply:
For reference, I'm working on pulling out the current simplified smoother from Smac and making a new plugin that would be able to support other classes of paths where we don't care about the curvature constraints, if that changes your view on this at all.

Reply:
I see no problem with the curvature constraint - it can be set to zero for "infeasible" planner inputs (or `minimum_turning_radius` can be set to zero with the same effect). I can add a test to check for an edge scenario (path with positions 0,0 -> 1,0 -> 0,0) whether some harmful `Inf`s or `NaN`s are produced, but even if it happened, it would be an easy-to-fix issue, so I think that narrowing the smoother to feasible inputs only is not required (a note about setting the parameters to zero can be added to docs).

A more difficult problem (however solvable - I have ideas for that but that's for another PR) would be an input path with multiple poses at one location (like the inplace rotations generated by diff-drive/omni `SmacPlannerLattice` planners). This can be mentioned in docs until it is resolved.

Reply:
However the 2D input was just an example, even with a feasible path producing planner you don't always need to finish with a specific orientation. The smoother can be used to take care of that and smooth out the final rotation maneuver.

Reply:
OK - we can put a pin in this item, I don't think it needs to be changed, I just need to keep the note here so I can go back through and really understand this in detail.